### PR TITLE
Merge `tft-model` into `master`

### LINKED ISSUE #225: TFT native implementation
### 🚀 Feature Request

Add native implementation of TFT model in the library

### Proposal

1. Read the [paper](https://arxiv.org/pdf/1912.09363.pdf)
2. Implement `TFTNativeModel`, see [DeepARNativeModel ](https://github.com/etna-team/etna/tree/master/etna/models/nn/deepar_native) as an example
  - `TFTNativeNet(DeepBaseNet)` -- implement the network itself as [LightningModule](https://lightning.ai/docs/pytorch/stable/common/lightning_module.html)
  - `TFTNativeModel(DeepBaseModel)` -- implement etna model wrapping the network, implement`params_to_tune`
  - Store complex blocks of the network, e.g. `GRN`, in additional `layers.py` file
  - Store `QuantileLoss` in additional `loss.py` file
  - Reference implementations: [PytorchForecasting](https://github.com/jdb78/pytorch-forecasting/tree/master/pytorch_forecasting/models/temporal_fusion_transformer), [GluonTS](https://github.com/awslabs/gluonts/tree/dev/src/gluonts/torch/model/tft), [Darts](https://github.com/unit8co/darts/blob/master/darts/models/forecasting/tft_model.py) + implementations from [paperswithcode](https://paperswithcode.com/paper/temporal-fusion-transformers-for)
4.  Try to reproduce metrics from the paper, see similar experiments for [DeepARNative](https://gitlab.tcsbank.ru/etna/experiments/-/tree/master/experiments/benchmark_deepar/directory)
- Find datasets from the paper(see etna-datasets)
- Run experiment with configuration from the paper
- Run the same experiment with implementation from PytorchForecasting
- Add metric from your experiments and from the paper into PR 
5. Update example [notebook](https://github.com/etna-team/etna/blob/master/examples/202-NN_examples.ipynb)

### Test cases

- Test `params_to_tune` similar to the other models
- Add sanity checks similar to the other models

### Additional context

We want to replace current implementation from PytorchForecasting without bumping the major version, we should somehow warn users that new implementation might work slightly different

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

## Need to be added:
1. Support of `prediction_intervals`
2. Support of `return_components`
3. Replace nn.MultiheadAttention with own implementation of InterpretableMultiHeadAttention, see [paper](https://arxiv.org/pdf/1912.09363.pdf)

## Need to be fixed:
1. Generation of samples in base class (now - high memory consumption)

## Closing issues
closes #225

### CODE REVIEW

**Comment by egoriyaa** [comment]
Need to be added:
1. Support of `prediction_intervals`
2. Support of `return_components`
3. Replace nn.MultiheadAttention with own implementation of InterpretableMultiHeadAttention, see [paper](https://arxiv.org/pdf/1912.09363.pdf)

Need to be fixed:
1. Generation of samples in base class (now - high memory consumption)

File: etna/models/nn/tft_native/tft.py [hunk 1]
```diff
@@ -0,0 +1,643 @@
...
+    def configure_optimizers(self) -> "torch.optim.Optimizer":
+        """Optimizer configuration."""
+        optimizer = torch.optim.Adam(self.parameters(), lr=self.lr, **self.optimizer_params)
+        return optimizer
+
+
+class TFTNativeModel(DeepBaseModel):
+    """TFT model. For more details read the `paper <https://arxiv.org/abs/1912.09363>`_.
+
+    Model needs label encoded inputs, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`.
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/tft.py:483]: Which values exactly should be label encoded? `static_categoricals` or smth else?

**Comment by alex-hse-repository** [comment]
> Need to be added:
> 
> 1. Support of `prediction_intervals`
> 2. Support of `return_components`
> 3. Replace nn.MultiheadAttention with own implementation of InterpretableMultiHeadAttention, see [paper](https://arxiv.org/pdf/1912.09363.pdf)
> 
> Need to be fixed:
> 
> 1. Generation of samples in base class (now - high memory consumption)

May you add it in PR description

File: etna/models/nn/tft_native/tft.py (hunk 1 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:483]: all categoricals - `static_categoricals`,  `time_varying_categoricals_encoder`,  `time_varying_categoricals_decoder`

File: etna/models/nn/tft_native/tft.py [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+        num_layers:
+            number of layers in LSTM layer
+        dropout:
+            dropout rate
+        hidden_size:
+            size of the hidden state
+        lr:
+            learning rate
+        static_categoricals:
+            categorical features for the whole series, e.g. `segment`
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:84]: static - there is one unique feature value for the whole series.
we don't have validation

[the remark this answers is not in the document]
Comment (d-a-bunin) [reply etna/models/nn/tft_native/tft.py:84]: I think we should describe it in our documentation. It isn't obvious for the user what does it mean "static" and is it checked or not (and how it behaves if it isn't really static).

File: etna/models/nn/tft_native/tft.py (hunk 1 above)
Comment (d-a-bunin) [reply etna/models/nn/tft_native/tft.py]: I think we need to clarify this in the docs.

File: etna/models/nn/tft_native/tft.py [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+        return optimizer
+
+
+class TFTNativeModel(DeepBaseModel):
+    """TFT model. For more details read the `paper <https://arxiv.org/abs/1912.09363>`_.
+
+    Model needs label encoded inputs for categorical features, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`.
+    Feature values that were not seen during `fit` should be set to NaN for expected behaviour with `strategy="none"`
+
+    Passed features for each feature type are not validated.
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/tft.py:486]: This is difficult to read and isn't specific at all. We should write smth more specific like: `Passed feature values aren't validated on being static or being correctly label encoded.`.

File: tests/test_models/test_inference/test_forecast.py [hunk 4]
```diff
@@ -1115,6 +1151,22 @@ def test_forecast_out_sample_suffix_failed_deepar(self, model, transforms, datas
...
+        [
+            (
+                TFTNativeModel(encoder_length=7, decoder_length=7, trainer_params=dict(max_epochs=1)),
+                [],
+                "example_tsds",
+            ),
+        ],
+    )
+    def test_forecast_out_sample_suffix_failed_tft(self, model, transforms, dataset_name, request):
+        """This test is expected to fail due to dependence of future predictions on previous predictions in TFT."""
```
Comment (alex-hse-repository) [inline tests/test_models/test_inference/test_forecast.py:1165]: Should it actually fail? TFT is "direct" model -- future predictions don't depend on previous predictions

File: tests/test_models/test_nn/tft_native/test_tft_native.py [hunk 5]
```diff
@@ -15,6 +16,20 @@
 from tests.test_models.utils import assert_sampling_is_valid
 
 
+def ts_label_encode(ts: TSDataset) -> TSDataset:
+    seg = SegmentEncoderTransform()
+    label1 = LabelEncoderTransform(in_column="categ_exog", out_column="categ_exog_label", strategy="none")
+    label2 = LabelEncoderTransform(in_column="categ_regr", out_column="categ_regr_label", strategy="none")
+    label3 = LabelEncoderTransform(in_column="categ_regr_new", out_column="categ_regr_new_label", strategy="none")
+
+    seg.fit_transform(ts)
```
Comment (alex-hse-repository) [inline tests/test_models/test_nn/tft_native/test_tft_native.py:25]: Why not just
```
ts.fit_transform([seg, label1, label2, label3])
```

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: tests/test_models/test_inference/test_forecast.py (hunk 4 above)
Comment (egoriyaa) [reply tests/test_models/test_inference/test_forecast.py:1165]: May be it's not clearly written. It is not autoregressive, but in these cases model uses different contexts

File: etna/models/nn/tft_native/tft.py [hunk 6]
```diff
@@ -195,7 +195,6 @@ def __init__(
 
         self.temporal_fusion_decoder = TemporalFusionDecoder(
             input_size=self.hidden_size,
-            decoder_length=self.decoder_length,
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py]: Initially model forecasted only for `self.decoder_length` steps.
Now model is forecasting for `horizon` steps
In all inference tests we are forecasting for horizon steps, where `horizon != decoder_length`

[the remark this answers is not in the document]
Comment (alex-hse-repository) [reply etna/models/nn/tft_native/tft.py]: May you add 3 cases for `decoder_length` and `horizon` (<, >, =) to `test_tft`

**Review by alex-hse-repository — APPROVED** [review]

File: etna/models/nn/tft_native/tft.py [hunk 7]
```diff
@@ -0,0 +1,647 @@
...
+            test_dataloader_params=test_dataloader_params,
+            val_dataloader_params=val_dataloader_params,
+            trainer_params=trainer_params,
+            split_params=split_params,
+        )
+
+    def params_to_tune(self) -> Dict[str, BaseDistribution]:
+        """Get default grid for tuning hyperparameters.
+
+        This grid tunes parameters: ``num_layers``, ``n_heads``, ``hidden_size``, ``lr``, ``dropout``, ``train_batch_size``.
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/tft.py:633]: It seems like it doesn't really tune `train_batch_size`.

**Review by d-a-bunin — APPROVED** [review]

**Review by alex-hse-repository — APPROVED** [review]
